### ISSUE TITLE
Possible inconsistency between paper and implementation regarding loss

### ISSUE THREAD

**Issue by iamzkf** [Westlake-AI/SemiReward#9]
1. **Observation of Inconsistency**  
   In the paper, the loss functions for the rewarder and generator are defined separately to ensure that:
   - When computing the generator loss, the rewarder does not receive gradients.
   - When computing the rewarder loss, the generator does not receive gradients.

   Specifically, the losses are defined as:

<img width="486" alt="Image" src="https://github.com/user-attachments/assets/0e1c002f-27ed-4ad4-a305-f7d2a138de58" />

   where the notation $\bar{G}$ and $\bar{R}$ indicates that the corresponding module's output is detached, so that no gradient flows back to it.

2. **Issue in the Current Implementation**  
   Take `semilearn/algorithms/srpseudolabel/srpseudolabel.py` line 180-190 as an example.
   In the code, the loss computation and backpropagation are performed as follows:
   ```python
   generator_loss = self.criterion(reward, torch.ones_like(reward).cuda(self.gpu))
   rewarder_loss = self.criterion(reward, cosine_similarity_score)

   self.generator_optimizer.zero_grad()
   self.rewarder_optimizer.zero_grad()

   generator_loss.backward(retain_graph=True)
   rewarder_loss.backward(retain_graph=True)

   self.generator_optimizer.step()
   self.rewarder_optimizer.step()
   ```
   This implementation appears to allow gradients from `generator_loss` to flow through the rewarder and vice versa, which contradicts the paper's explanation that each loss should be computed without affecting the other module.

3. **Request for Clarification**  
   Could you please clarify whether this is the intended behavior or if there might be an issue in the implementation? I am not sure if my understanding is wrong or if the code indeed deviates from the formulation described in the paper.

Thank you very much for your time and for making this work available.
